### PR TITLE
Store providers tokens as secrets

### PR DESCRIPTION
**Pull request by dramich**
Problem:
Provider tokens are stored on the principal objects in k8s

Solution:
Move provider tokens to be a k8s secret

Issue https://github.com/rancherlabs/the-ranch/issues/134

### CODE REVIEW

File: pkg/auth/providers/common/usermanager.go [hunk 1]
```diff
@@ -232,13 +232,13 @@ func (m *userManager) EnsureUser(principalName, displayName string) (*v3.User, e
 		PrincipalIDs: []string{principalName},
 	}
 
-	logrus.Infof("Creating user for principal %v", principalName)
+	logrus.Infof("Creating user for principal %s", principalName)
```
Comment (cjellick) [inline pkg/auth/providers/common/usermanager.go]: back to %v

File: pkg/auth/providers/github/github_provider.go [hunk 2]
```diff
@@ -227,8 +227,10 @@ func (g *ghProvider) GetPrincipal(principalID string, token v3.Token) (v3.Princi
 		return v3.Principal{}, err
 	}
 
-	accessToken := token.ProviderInfo["access_token"]
-
+	accessToken, err := g.tokenMGR.GetSecret(&token)
+	if err != nil {
+		return v3.Principal{}, err
+	}
```
Comment (cjellick) [inline pkg/auth/providers/github/github_provider.go]: add fallback logic

Comment (cjellick) [reply pkg/auth/providers/github/github_provider.go]: on azure too

File: pkg/auth/tokens/manager.go [hunk 3]
```diff
@@ -59,6 +64,8 @@ type Manager struct {
 	userIndexer         cache.Indexer
 	tokenIndexer        cache.Indexer
 	userLister          v3.UserLister
+	secretClient        v1.SecretInterface
+	secretCache         v1.SecretLister
```
Comment (cjellick) [inline pkg/auth/tokens/manager.go]: secrets and secretLIister please

File: pkg/auth/tokens/manager.go [hunk 4]
```diff
@@ -224,7 +231,7 @@ func (m *Manager) deleteToken(tokenAuthValue string) (int, error) {
 func (m *Manager) deleteTokenByName(tokenName string) (int, error) {
 	err := m.tokensClient.Delete(tokenName, &metav1.DeleteOptions{})
 	if err != nil {
-		if e2, ok := err.(*errors.StatusError); ok && e2.Status().Code == 404 {
+		if e2, ok := err.(*apierrors.StatusError); ok && e2.Status().Code == 404 {
```
Comment (cjellick) [inline pkg/auth/tokens/manager.go]: apierrors.IsNotFound()

File: pkg/auth/tokens/manager.go [hunk 5]
```diff
@@ -272,12 +279,14 @@ func (m *Manager) deriveToken(request *types.APIContext) error {
 	bytes, err := ioutil.ReadAll(r.Body)
 	if err != nil {
 		logrus.Errorf("GetToken failed with error: %v", err)
+		return httperror.NewAPIError(httperror.InvalidBodyContent, fmt.Sprintf("%s", err))
```
Comment (cjellick) [inline pkg/auth/tokens/manager.go]: dorp the error log and just return the error

File: pkg/auth/tokens/manager.go [hunk 6]
```diff
@@ -446,6 +455,55 @@ func (m *Manager) removeToken(request *types.APIContext) error {
...
+	}
+
+	return string(cachedSecret.Data[token.AuthProvider]), nil
+}
+
+func (m *Manager) UpdateSecret(userID, provider, secret string) error {
+	cachedSecret, err := m.secretCache.Get(secretNamespace, userID+secretNameEnding)
+	if err != nil {
+		return err
+	}
```
Comment (cjellick) [inline pkg/auth/tokens/manager.go]: cachedSecret = cachedSecret.DeepCopy()

File: pkg/auth/providers/common/usermanager.go (hunk 1 above)
Comment (dramich) [reply pkg/auth/providers/common/usermanager.go]: done

File: pkg/auth/providers/github/github_provider.go (hunk 2 above)
Comment (dramich) [reply pkg/auth/providers/github/github_provider.go]: Done

File: pkg/auth/tokens/manager.go (hunk 3 above)
Comment (dramich) [reply pkg/auth/tokens/manager.go]: done

File: pkg/auth/tokens/manager.go (hunk 4 above)
Comment (dramich) [reply pkg/auth/tokens/manager.go]: done

File: pkg/auth/tokens/manager.go (hunk 5 above)
Comment (dramich) [reply pkg/auth/tokens/manager.go]: done

File: pkg/auth/tokens/manager.go (hunk 6 above)
Comment (dramich) [reply pkg/auth/tokens/manager.go]: done

File: pkg/auth/tokens/manager.go [hunk 7]
```diff
@@ -508,8 +566,15 @@ var uaBackoff = wait.Backoff{
 	Steps:    5,
 }
 
-func (m *Manager) NewLoginToken(userID string, userPrincipal v3.Principal, groupPrincipals []v3.Principal, providerInfo map[string]string, ttl int64, description string) (v3.Token, error) {
-	provider := getAuthProviderName(userPrincipal.Name)
+func (m *Manager) NewLoginToken(userID string, userPrincipal v3.Principal, groupPrincipals []v3.Principal, providerToken string, ttl int64, description string) (v3.Token, error) {
+	provider := userPrincipal.Provider
+
+	if provider == "github" || provider == "azuread" && providerToken != "" {
```
[the remark this answers is not in the document]
Comment (dramich) [reply pkg/auth/tokens/manager.go]: done

File: pkg/controllers/management/auth/user.go [hunk 8]
```diff
@@ -354,3 +363,14 @@ func (l *userLifecycle) deleteUserNamespace(username string) error {
 
 	return nil
 }
+
+func (l *userLifecycle) deleteUserSecret(username string) error {
+	_, err := l.secretsLister.Get("cattle-system", username)
```
[the remark this answers is not in the document]
Comment (dramich) [reply pkg/controllers/management/auth/user.go]: done

**Comment by cjellick** [comment]
LGTM (assuming green build)

**Pull request merged**